File: WireExtensionComponents/Views/AvatarImageView.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


#import "AvatarImageView.h"
#import <WireExtensionComponents/WireExtensionComponents-Swift.h>

@interface AvatarImageView ()

@property (nonatomic) RoundedView *containerView;
@property (nonatomic) UIImageView *imageView;
@property (nonatomic) UILabel *initials;

@end

@implementation AvatarImageView

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];

    if (self) {
        [self setup];
    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder
{
    self = [super initWithCoder:aDecoder];
    if (self) {
        [self setup];
    }
    return self;
}

- (BOOL)isOpaque
{
    return NO;
}

- (void)setup
{
    _shape = AvatarImageViewShapeCircle;
    [self createContainerView];
    [self createImageView];
    [self createInitials];

    self.showInitials = YES;
    [self createConstraints];
    [self updateCornerRadius];
}

- (void)setShape:(AvatarImageViewShape)shape
{
    _shape = shape;
    [self updateCornerRadius];
}

- (void)updateCornerRadius
{
    switch (self.shape) {
        case AvatarImageViewShapeRectangle: {
            [self.containerView toggleRectangle];
            return;
        }
        case AvatarImageViewShapeCircle: {
            [self.containerView toggleCircle];
            return;
        }
        case AvatarImageViewShapeRoundedRelative: {
            [self.containerView setRelativeCornerRadiusWithMultiplier:(CGFloat)1/6 dimension:MaskDimensionHeight];
            return;
        }
    }
}

- (void)createContainerView
{
    self.containerView = [RoundedView new];
    self.containerView.clipsToBounds = YES;
    [self addSubview:self.containerView];
}

- (void)createImageView
{
    self.imageView = [UIImageView new];
    self.imageView.layer.borderColor = [UIColor clearColor].CGColor;
    self.imageView.opaque = NO;
    [self.containerView addSubview:self.imageView];
}

- (void)createInitials
{
    self.initials = [UILabel new];
    self.initials.opaque = NO;
    [self.containerView addSubview:self.initials];
}

- (void)createConstraints
{
    self.containerView.translatesAutoresizingMaskIntoConstraints = NO;
    self.imageView.translatesAutoresizingMaskIntoConstraints = NO;
    self.initials.translatesAutoresizingMaskIntoConstraints = NO;

    NSArray<NSLayoutConstraint *> *constraints =
    @[
      // containerView
      [self.containerView.leadingAnchor constraintEqualToAnchor:self.leadingAnchor],
      [self.containerView.topAnchor constraintEqualToAnchor:self.topAnchor],
      [self.containerView.trailingAnchor constraintEqualToAnchor:self.trailingAnchor],
      [self.containerView.bottomAnchor constraintEqualToAnchor:self.bottomAnchor],
      [self.containerView.widthAnchor constraintEqualToAnchor:self.containerView.heightAnchor],

      // imageView
      [self.imageView.leadingAnchor constraintEqualToAnchor:self.containerView.leadingAnchor],
      [self.imageView.topAnchor constraintEqualToAnchor:self.containerView.topAnchor],
      [self.imageView.trailingAnchor constraintEqualToAnchor:self.containerView.trailingAnchor],
      [self.imageView.bottomAnchor constraintEqualToAnchor:self.containerView.bottomAnchor],

      // initials
      [self.initials.centerXAnchor constraintEqualToAnchor:self.containerView.centerXAnchor],
      [self.initials.centerYAnchor constraintEqualToAnchor:self.containerView.centerYAnchor]
      ];

    [NSLayoutConstraint activateConstraints:constraints];
}

- (BOOL)showInitials
{
    return self.initials.isHidden;
}

- (void)setShowInitials:(BOOL)showInitials
{
    self.initials.hidden = !showInitials;
}

@end

File: Wire-iOS/Sources/UserInterface/Components/AvatarImageView.swift
//
// Wire
// Copyright (C) 2018 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import UIKit

/**
 * A view that displays the avatar of a user, either as text initials or as an image.
 */

class AvatarImageView: UIControl {

    /**
     * The different, mutually-exclusive forms of avatars
     */

    enum Avatar: Equatable {
        case image(UIImage)
        case text(String)
    }

    /**
     * The different shapes of avatars.
     */

    enum Shape {
        case rectangle, circle, relative
    }

    // MARK: - Properties

    /// The avatar to display.
    var avatar: Avatar? {
        didSet {
            if avatar != oldValue {
                updateAvatar()
            }
        }
    }

    /// The shape of the avatar
    var shape: Shape = .circle {
        didSet {
            if shape != oldValue {
                updateShape()
            }
        }
    }

    /// Whether to allow initials.
    var allowsInitials: Bool = true {
        didSet {
            if allowsInitials != oldValue {
                updateAvatar()
            }
        }
    }

    /// The background color for the image.
    var imageBackgroundColor: UIColor? {
        get { return container.backgroundColor }
        set {
            container.backgroundColor = newValue
        }
    }

    /// The font to use of the initials label.
    var initialsFont: UIFont {
        get { return initialsLabel.font }
        set { initialsLabel.font = newValue }
    }

    /// The color to use for the initials label.
    var initialsColor: UIColor {
        get { return initialsLabel.textColor }
        set { initialsLabel.textColor = newValue }
    }

    override var contentMode: UIView.ContentMode {
        didSet {
            if contentMode != oldValue {
                imageView.contentMode = contentMode
            }
        }
    }

    /// The view that contains the avatar.
    var container = RoundedView()

    private let imageView = UIImageView()
    private let initialsLabel = UILabel()

    // MARK: - Initialization

    override init(frame: CGRect) {
        super.init(frame: frame)
        configureSubviews()
        configureConstraints()
    }

    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        configureSubviews()
        configureConstraints()
    }

    private func configureSubviews() {
        contentMode = .scaleAspectFill
        
        isOpaque = false
        imageView.isOpaque = false
        container.backgroundColor = .yellow

        initialsLabel.textColor = .white
        initialsLabel.font = .systemFont(ofSize: 17)

        container.clipsToBounds = true
        container.isUserInteractionEnabled = false

        imageView.isAccessibilityElement = false
        initialsLabel.isAccessibilityElement = false

        addSubview(container)
        container.addSubview(imageView)
        container.addSubview(initialsLabel)

        updateAvatar()
        updateShape()
    }

    private func configureConstraints() {
        container.translatesAutoresizingMaskIntoConstraints = false
        imageView.translatesAutoresizingMaskIntoConstraints = false
        initialsLabel.translatesAutoresizingMaskIntoConstraints = false

        NSLayoutConstraint.activate([
            // containerView
            container.leadingAnchor.constraint(equalTo: leadingAnchor),
            container.topAnchor.constraint(equalTo: topAnchor),
            container.trailingAnchor.constraint(equalTo: trailingAnchor),
            container.bottomAnchor.constraint(equalTo: bottomAnchor),
            container.widthAnchor.constraint(equalTo: container.heightAnchor),

            // imageView
            imageView.leadingAnchor.constraint(equalTo: container.leadingAnchor),
            imageView.topAnchor.constraint(equalTo: container.topAnchor),
            imageView.trailingAnchor.constraint(equalTo: container.trailingAnchor),
            imageView.bottomAnchor.constraint(equalTo: container.bottomAnchor),

            // initials
            initialsLabel.centerXAnchor.constraint(equalTo: container.centerXAnchor),
            initialsLabel.centerYAnchor.constraint(equalTo: container.centerYAnchor)
        ])
    }

    /**
     * Updates the image constraints hugging and resistance priorities.
     * - parameter resistance: The compression resistance priority.
     * - parameter hugging: The content hugging priority.
     */

    func setImageConstraint(resistance: Float, hugging: Float) {
        imageView.setContentHuggingPriority(UILayoutPriority(rawValue: hugging), for: .vertical)
        imageView.setContentHuggingPriority(UILayoutPriority(rawValue: hugging), for: .horizontal)
        imageView.setContentCompressionResistancePriority(UILayoutPriority(rawValue: resistance), for: .vertical)
        imageView.setContentCompressionResistancePriority(UILayoutPriority(rawValue: resistance), for: .horizontal)
    }

    // MARK: - Content

    /// Updates the displayed avatar.
    private func updateAvatar() {
        switch avatar {
        case .image(let image)?:
            imageView.image = image
            initialsLabel.text = nil
            imageView.isHidden = false
            initialsLabel.isHidden = true

        case .text(let text)?:
            guard allowsInitials else { fallthrough }

            imageView.image = nil
            initialsLabel.text = text
            imageView.isHidden = true
            initialsLabel.isHidden = false

        case .none:
            imageView.image = nil
            initialsLabel.text = nil
            imageView.isHidden = true
            initialsLabel.isHidden = true
        }
    }

    /// Updates the shape of the displayed avatar.
    private func updateShape() {
        switch shape {
        case .circle:
            container.shape = .circle
        case .rectangle:
            container.shape = .rectangle
        case .relative:
            container.shape = .relative(multiplier: 1/6, dimension: .height)
        }
    }

}
